### PR TITLE
feat(Core/Spell): Implement ValidateSpellInfo

### PR DESCRIPTION
**Pull request by Kitzunu**
* cherry-pick from https://github.com/TrinityCore/TrinityCore/commit/2b5d7eef3a091e67c27e853d43ac0bebbfa67531 & https://github.com/TrinityCore/TrinityCore/commit/a0a158b5b851db7e2c16819ec89e913d914a3aba

## Changes Proposed:
-  Implement ValidateSpellInfo for bool Validate() 
```cpp
// This allows to use it this way, short and concise 
        bool Validate(SpellInfo const* /*spellInfo*/) override
        {
            return ValidateSpellInfo({ SPELL_ID });
        }

// Instead of this way
        bool Validate(SpellInfo const* /*spellInfo*/) override
        {
            if (!sSpellMgr->GetSpellInfo(SPELL_ID))
                return false;
            return true;
        }
```

## Tests Performed:
- Builds
- tested in-game

```
2021-01-22 00:05:52 Validating spell scripts...
2021-01-22 00:05:52 >> Validated 2218 scripts in 5 ms
```

## How to Test the Changes:

1. Test any spellscript, I personally tested it with `spell_item_instant_statue`
2. First test with `return ValidateSpellInfo({ SPELL_ID });` to see that it works
3. then test with `return !ValidateSpellInfo({ SPELL_ID });` and see that you will get this error while starting WorldServer as intended
```
ERROR: TSCR: Spell `75731` did not pass Validate() function of script `spell_item_instant_statue` - script will be not added to the spell
```

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by pklloveyou** [comment]
I put this on the server this morning to see if players will report errors

**Comment by Kitzunu** [comment]
Any updates?

**Comment by pklloveyou** [comment]
So far, no player has reported an error

**Comment by pklloveyou** [comment]
So far, no player has reported an error，

File: src/server/scripts/Spells/spell_item.cpp [hunk 1]
```diff
@@ -3590,7 +3591,9 @@ class spell_item_demon_broiled_surprise : public SpellScriptLoader
 
         bool Validate(SpellInfo const* /*spell*/) override
         {
-            if (!sSpellMgr->GetSpellInfo(SPELL_CREATE_DEMON_BROILED_SURPRISE) || !sObjectMgr->GetCreatureTemplate(NPC_ABYSSAL_FLAMEBRINGER) || !sObjectMgr->GetQuestTemplate(QUEST_SUPER_HOT_STEW))
+            return ValidateSpellInfo({ SPELL_CREATE_DEMON_BROILED_SURPRISE });
+
+            if (!sObjectMgr->GetCreatureTemplate(NPC_ABYSSAL_FLAMEBRINGER) || !sObjectMgr->GetQuestTemplate(QUEST_SUPER_HOT_STEW))
                 return false;
             return true;
```
Comment (Helias) [inline src/server/scripts/Spells/spell_item.cpp:3598]: same here, check the code after the return

File: src/server/scripts/Spells/spell_item.cpp [hunk 2]
```diff
@@ -4042,7 +4039,9 @@ class spell_item_chicken_cover : public SpellScriptLoader
 
         bool Validate(SpellInfo const* /*spell*/) override
         {
-            if (!sSpellMgr->GetSpellInfo(SPELL_CHICKEN_NET) || !sSpellMgr->GetSpellInfo(SPELL_CAPTURE_CHICKEN_ESCAPE) || !sObjectMgr->GetQuestTemplate(QUEST_CHICKEN_PARTY) || !sObjectMgr->GetQuestTemplate(QUEST_FLOWN_THE_COOP))
+            return ValidateSpellInfo({ SPELL_CHICKEN_NET, SPELL_CAPTURE_CHICKEN_ESCAPE });
+
+            if (!sObjectMgr->GetQuestTemplate(QUEST_CHICKEN_PARTY) || !sObjectMgr->GetQuestTemplate(QUEST_FLOWN_THE_COOP))
                 return false;
             return true;
         }
```
Comment (Helias) [inline src/server/scripts/Spells/spell_item.cpp:4047]: sam here

**Review by Helias — CHANGES_REQUESTED** [review]

File: src/server/scripts/Spells/spell_dk.cpp [hunk 3]
```diff
@@ -1780,8 +1769,9 @@ class spell_dk_ghoul_explode : public SpellScriptLoader
 
         bool Validate(SpellInfo const* spellInfo) override
         {
-            if (!sSpellMgr->GetSpellInfo(SPELL_DK_CORPSE_EXPLOSION_TRIGGERED)
-                    || spellInfo->Effects[EFFECT_2].CalcValue() <= 0)
+            return ValidateSpellInfo({ SPELL_DK_CORPSE_EXPLOSION_TRIGGERED });
+
+            if (spellInfo->Effects[EFFECT_2].CalcValue() <= 0)
                 return false;
             return true;
```
[the remark this answers is not in the document]
Comment (Kitzunu) [reply src/server/scripts/Spells/spell_dk.cpp:1776]: Yea I spoke with a friemd and he says checking calc value is useless. But I wanted to leave it for someone to comment on it.

Thoughts?

File: src/server/scripts/Spells/spell_item.cpp (hunk 1 above)
Comment (Kitzunu) [reply src/server/scripts/Spells/spell_item.cpp:3598]: I  will remove this, validate is just for spells. This other things doesnt matter

File: src/server/game/Spells/SpellScript.cpp [hunk 4]
```diff
@@ -14,12 +14,28 @@ bool _SpellScript::_Validate(SpellInfo const* entry)
...
+        {
+            sLog->outError("_SpellScript::_ValidateSpellInfo: Spell %u does not exist.", *begin);
+            allValid = false;
+        }
+
+        ++begin;
+    }
+    return allValid;
+}
+
```
[the remark this answers is not in the document]
Comment (Kitzunu) [reply src/server/game/Spells/SpellScript.cpp:38]: I don't know... Shauren put it there for a reason, but I am too much of a noob to know why https://github.com/TrinityCore/TrinityCore/commit/a0a158b5b851db7e2c16819ec89e913d914a3aba#diff-8ef12e6a47df19676017ee28acdfff0754e9438967a2d57e1dd46a71b3c9cacdR39

[the remark this answers is not in the document]
Comment (Helias) [reply src/server/game/Spells/SpellScript.cpp:38]: the only reason that I can see is to print in the log all the spells, we could leave as it is for the log.

**Review by Helias — APPROVED** [review]

**Comment by pklloveyou** [comment]
Tested

**Comment by LannyE** [comment]
I just updated my core -> 3.0 -> latest master which includes this change.  Now I see this:
image

Is anyone else seeing this?  Do I need to make an issue report for it?

**Comment by Kitzunu** [comment]
Not caused by this PR, I have not touched any bwl scripts + the CI passes.

Building lastest master atm to see if I have it as well

**Comment by Kitzunu** [comment]
image
Apart from me not updating my DB (So these errors are OK!), I do not get your error.

Perhaps your DBC is faulty

**Comment by LannyE** [comment]
I think I found where that spell was changed/added #3673.  It was merged 1-31-21.   I'll post there.  Thanks for checking.

**Pull request merged**